File: web/tests/reference/dynamic-programming/house-robber-iii.js
function rob(root) {
  const walk = (node) => {
    if (node === null) return [0, 0];
    const left = walk(node.left);
    const right = walk(node.right);
    const withNode = node.val + left[1] + right[1];
    const without = Math.max(...left) + Math.max(...right);
    return [withNode, without];
  };
  return Math.max(...walk(root));
}

File: web/tests/reference/dynamic-programming/house-robber-iii.py
def rob(root):
    def walk(node):
        if node is None:
            return (0, 0)
        left = walk(node.left)
        right = walk(node.right)
        with_node = node.val + left[1] + right[1]
        without = max(left) + max(right)
        return (with_node, without)

    return max(walk(root))
